lib/src/parallel_prefix_operations.dart

```diff
@@ -0,0 +1,250 @@
+      : super(inps, 'ripple') {
+    final iseq = <Logic>[];
+
+    inps.forEachIndexed((i, el) {
+      iseq.add(addInput('i$i', el, width: el.width));
+      _oseq.add(addOutput('o$i', width: el.width));
+    });
+
+    for (var i = 0; i < iseq.length; ++i) {
+      if (i == 0) {
+        _oseq[i] <= iseq[i];
+      } else {
+        _oseq[i] <= op(_oseq[i - 1], iseq[i]);
+      }
+    }
+  }
+}
+
+/// Sklansky shaped pp tree
+class Sklansky extends ParallelPrefix {
+  /// Sklansky constructor
+  Sklansky(List<Logic> inps, Logic Function(Logic, Logic) op)
+      : super(inps, 'sklansky') {
+    final iseq = <Logic>[];
+
+    inps.forEachIndexed((i, el) {
+      iseq.add(addInput('i$i', el, width: el.width));
+      _oseq.add(addOutput('o$i', width: el.width));
+    });
+
+    if (iseq.length == 1) {
+      _oseq[0] <= iseq[0];
+    } else {
+      final n = iseq.length;
+      final m = largestPow2LessThan(n);
+      final u = Sklansky(iseq.getRange(0, m).toList(), op).val;
+      final v = Sklansky(iseq.getRange(m, n).toList(), op).val;
+      u.forEachIndexed((i, el) {
+        _oseq[i] <= el;
+      });
+      v.forEachIndexed((i, el) {
+        _oseq[m + i] <= op(u[m - 1], el);
+      });
+    }
+  }
+}
+
+/// KoggeStone shaped pp tree
+class KoggeStone extends ParallelPrefix {
+  /// KoggeStone constructor
+  KoggeStone(List<Logic> inps, Logic Function(Logic, Logic) op)
+      : super(inps, 'kogge_stone') {
+    final iseq = <Logic>[];
+
+    inps.forEachIndexed((i, el) {
+      iseq.add(addInput('i$i', el, width: el.width));
+      _oseq.add(addOutput('o$i', width: el.width));
+    });
+
+    var skip = 1;
+
+    while (skip < inps.length) {
+      for (var i = inps.length - 1; i >= skip; --i) {
+        iseq[i] = op(iseq[i - skip], iseq[i]);
+      }
+      skip *= 2;
+    }
+
+    iseq.forEachIndexed((i, el) {
+      _oseq[i] <= el;
+    });
+  }
+}
+
+/// BrentKung shaped pp tree
+class BrentKung extends ParallelPrefix {
+  /// BrentKung constructor
+  BrentKung(List<Logic> inps, Logic Function(Logic, Logic) op)
+      : super(inps, 'brent_kung') {
+    final iseq = <Logic>[];
+
+    inps.forEachIndexed((i, el) {
+      iseq.add(addInput('i$i', el, width: el.width));
+      _oseq.add(addOutput('o$i', width: el.width));
+    });
+
+    // Reduce phase
+    var skip = 2;
+    while (skip <= inps.length) {
+      for (var i = skip - 1; i < inps.length; i += skip) {
+        iseq[i] = op(iseq[i - skip ~/ 2], iseq[i]);
+      }
+      skip *= 2;
+    }
+
+    // Prefix Phase
+    skip = largestPow2LessThan(inps.length);
+    while (skip > 2) {
+      for (var i = 3 * (skip ~/ 2) - 1; i < inps.length; i += skip) {
+        iseq[i] = op(iseq[i - skip ~/ 2], iseq[i]);
+      }
+      skip ~/= 2;
+    }
+
+    // Final row
+    for (var i = 2; i < inps.length; i += 2) {
+      iseq[i] = op(iseq[i - 1], iseq[i]);
+    }
+
+    iseq.forEachIndexed((i, el) {
+      _oseq[i] <= el;
+    });
+  }
+}
+
+/// Or scan based on pp tree
```

Comment:
In both comments and module names, can we write out `ParallelPrefix` instead of `PP` so that it's more clear to those unfamiliar?

Reply:
Sure.  But ParallelPrefixAdder is really long,  Should we do ParPrefixAdder?  Or stick with the full name.

Reply:
I've been trying to follow Dart style and design conventions where possible since they are well organized and thoughtful, and making a new set of conventions for ROHD specifically seems like a daunting task.

Dart recommends avoiding abbreviations where possible:
https://dart.dev/effective-dart/design#avoid-abbreviations

With autocomplete in modern IDEs, I'm inclined to agree with it, and I think it's worth the extra few letters for the clarity it provides as an API